File: weather.cs
// weather.cs - Погодное приложение на C# Windows Forms
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Net.Http;
using System.Threading.Tasks;
using System.Windows.Forms;
using Newtonsoft.Json.Linq;

namespace WeatherApp
{
    public partial class MainForm : Form
    {
        private TextBox cityBox;
        private ComboBox unitsBox;
        private Label tempLabel, feelsLabel, humidityLabel, windLabel, pressureLabel, descLabel;
        private ListBox forecastList;
        private string apiKey;

        public MainForm()
        {
            apiKey = Environment.GetEnvironmentVariable("OPENWEATHER_API_KEY");
            if (string.IsNullOrEmpty(apiKey))
            {
                apiKey = Microsoft.VisualBasic.Interaction.InputBox("Введите OpenWeatherMap API ключ:", "API Key");
            }
            InitializeComponent();
        }

        private void InitializeComponent()
        {
            this.Text = "Погодное приложение C#";
            this.Size = new Size(600, 550);
            this.StartPosition = FormStartPosition.CenterScreen;
            
            var topPanel = new FlowLayoutPanel { Dock = DockStyle.Top, Padding = new Padding(5) };
            topPanel.Controls.Add(new Label { Text = "Город:", AutoSize = true });
            cityBox = new TextBox { Width = 200 };
            topPanel.Controls.Add(cityBox);
            var searchBtn = new Button { Text = "Поиск" };
            searchBtn.Click += async (s, e) => await FetchWeather();
            topPanel.Controls.Add(searchBtn);
            unitsBox = new ComboBox { DropDownStyle = ComboBoxStyle.DropDownList, Items = { "metric (°C)", "imperial (°F)" }, SelectedIndex = 0 };
            topPanel.Controls.Add(unitsBox);
            this.Controls.Add(topPanel);
            
            var infoPanel = new TableLayoutPanel { RowCount = 6, ColumnCount = 2, Dock = DockStyle.Top, Height = 200, Padding = new Padding(10) };
            infoPanel.Controls.Add(new Label { Text = "Температура:" }, 0,0);
            tempLabel = new Label(); infoPanel.Controls.Add(tempLabel, 1,0);
            infoPanel.Controls.Add(new Label { Text = "Ощущается:" }, 0,1);
            feelsLabel = new Label(); infoPanel.Controls.Add(feelsLabel, 1,1);
            infoPanel.Controls.Add(new Label { Text = "Влажность:" }, 0,2);
            humidityLabel = new Label(); infoPanel.Controls.Add(humidityLabel, 1,2);
            infoPanel.Controls.Add(new Label { Text = "Ветер:" }, 0,3);
            windLabel = new Label(); infoPanel.Controls.Add(windLabel, 1,3);
            infoPanel.Controls.Add(new Label { Text = "Давление:" }, 0,4);
            pressureLabel = new Label(); infoPanel.Controls.Add(pressureLabel, 1,4);
            infoPanel.Controls.Add(new Label { Text = "Описание:" }, 0,5);
            descLabel = new Label(); infoPanel.Controls.Add(descLabel, 1,5);
            this.Controls.Add(infoPanel);
            
            forecastList = new ListBox { Dock = DockStyle.Fill };
            this.Controls.Add(forecastList);
        }

        private async Task FetchWeather()
        {
            string city = cityBox.Text.Trim();
            if (string.IsNullOrEmpty(city)) return;
            string units = unitsBox.SelectedIndex == 0 ? "metric" : "imperial";
            using var client = new HttpClient();
            string url = $"https://api.openweathermap.org/data/2.5/weather?q={city}&appid={apiKey}&units={units}&lang=ru";
            var response = await client.GetAsync(url);
            if (!response.IsSuccessStatusCode)
            {
                MessageBox.Show("Город не найден или ошибка API");
                return;
            }
            var json = JObject.Parse(await response.Content.ReadAsStringAsync());
            DisplayCurrent(json, units);
            
            string forecastUrl = $"https://api.openweathermap.org/data/2.5/forecast?q={city}&appid={apiKey}&units={units}&lang=ru";
            var forecastResp = await client.GetAsync(forecastUrl);
            if (forecastResp.IsSuccessStatusCode)
            {
                var forecastJson = JObject.Parse(await forecastResp.Content.ReadAsStringAsync());
                DisplayForecast(forecastJson, units);
            }
        }

        private void DisplayCurrent(JObject json, string units)
        {
            var main = json["main"];
            double temp = (double)main["temp"];
            double feels = (double)main["feels_like"];
            int humidity = (int)main["humidity"];
            int pressure = (int)main["pressure"];
            double windSpeed = (double)json["wind"]["speed"];
            string desc = (string)json["weather"][0]["description"];
            string tempUnit = units == "metric" ? "°C" : "°F";
            string windUnit = units == "metric" ? "м/с" : "mph";
            tempLabel.Text = $"{temp:F1} {tempUnit}";
            feelsLabel.Text = $"{feels:F1} {tempUnit}";
            humidityLabel.Text = $"{humidity}%";
            windLabel.Text = $"{windSpeed:F1} {windUnit}";
            pressureLabel.Text = $"{pressure} гПа";
            descLabel.Text = desc;
        }

        private void DisplayForecast(JObject json, string units)
        {
            var list = json["list"];
            var days = new Dictionary<string, List<double>>();
            foreach (var item in list)
            {
                long dt = (long)item["dt"];
                var date = DateTimeOffset.FromUnixTimeSeconds(dt).DateTime;
                string day = date.ToString("yyyy-MM-dd");
                double temp = (double)item["main"]["temp"];
                if (!days.ContainsKey(day)) days[day] = new List<double>();
                days[day].Add(temp);
            }
            forecastList.Items.Clear();
            string tempUnit = units == "metric" ? "°C" : "°F";
            int count = 0;
            foreach (var day in days.Keys)
            {
                if (count++ >= 5) break;
                var temps = days[day];
                double min = temps.Min();
                double max = temps.Max();
                forecastList.Items.Add($"{day}: {min:F1} .. {max:F1} {tempUnit}");
            }
        }

        [STAThread]
        static void Main()
        {
            Application.EnableVisualStyles();
            Application.Run(new MainForm());
        }
    }
}

File: weather.py
"""
weather.py - Погодное приложение на Python (CLI + Tkinter опционально)
Для работы требуется requests и установленный API ключ в переменной окружения OPENWEATHER_API_KEY
"""

import os
import json
import sys
from datetime import datetime
import requests
from collections import defaultdict

API_KEY = os.environ.get("OPENWEATHER_API_KEY", "")
if not API_KEY:
    print("❌ Ошибка: установите переменную окружения OPENWEATHER_API_KEY")
    sys.exit(1)

BASE_URL = "https://api.openweathermap.org/data/2.5"

def get_weather(city, units="metric"):
    """Получить текущую погоду и прогноз"""
    # текущая погода
    current_url = f"{BASE_URL}/weather?q={city}&appid={API_KEY}&units={units}&lang=ru"
    current_resp = requests.get(current_url)
    if current_resp.status_code != 200:
        return None, None, current_resp.status_code
    current = current_resp.json()
    
    # прогноз на 5 дней
    forecast_url = f"{BASE_URL}/forecast?q={city}&appid={API_KEY}&units={units}&lang=ru"
    forecast_resp = requests.get(forecast_url)
    if forecast_resp.status_code != 200:
        forecast = None
    else:
        forecast = forecast_resp.json()
    return current, forecast, 200

def display_current(current, units):
    temp_unit = "°C" if units == "metric" else "°F"
    wind_unit = "m/s" if units == "metric" else "mph"
    temp = current['main']['temp']
    feels_like = current['main']['feels_like']
    humidity = current['main']['humidity']
    wind = current['wind']['speed']
    pressure = current['main']['pressure']
    weather_desc = current['weather'][0]['description']
    city = current['name']
    
    print(f"\n📍 {city}")
    print(f"🌡️ Температура: {temp}{temp_unit} (ощущается как {feels_like}{temp_unit})")
    print(f"💧 Влажность: {humidity}%")
    print(f"💨 Ветер: {wind} {wind_unit}")
    print(f"🔽 Давление: {pressure} гПа")
    print(f"☁️ {weather_desc.capitalize()}")

def display_forecast(forecast, units):
    if not forecast:
        print("\n⚠️ Прогноз на 5 дней недоступен")
        return
    temp_unit = "°C" if units == "metric" else "°F"
    # группировка по дням
    daily = defaultdict(list)
    for item in forecast['list']:
        dt = datetime.fromtimestamp(item['dt'])
        day = dt.strftime("%Y-%m-%d")
        daily[day].append(item['main']['temp'])
    
    print("\n📅 Прогноз на 5 дней (макс/мин):")
    for day, temps in list(daily.items())[:5]:
        max_temp = max(temps)
        min_temp = min(temps)
        print(f"{day}: {min_temp:.1f}{temp_unit} .. {max_temp:.1f}{temp_unit}")

def main():
    print("🌤️ Погодное приложение (Python)")
    units = input("Выберите единицы (1-метрические °C, 2-имперские °F): ").strip()
    units = "imperial" if units == "2" else "metric"
    while True:
        city = input("\nВведите название города (или 'q' для выхода): ").strip()
        if city.lower() == 'q':
            break
        current, forecast, status = get_weather(city, units)
        if status == 200:
            display_current(current, units)
            display_forecast(forecast, units)
        elif status == 401:
            print("Ошибка API: неверный ключ")
            break
        elif status == 404:
            print("Город не найден")
        else:
            print(f"Ошибка {status}")

if __name__ == "__main__":
    main()
